### DISCUSSION TITLE
efficiently getting point ids from a given cell

### DISCUSSION THREAD

**Discussion by firstkingofrome** [pyvista/pyvista#2838] [Q&A] [answered]
Hello, I was wondering if there is an easy way to get point ids from a given cell. For example:
```
from pyvista import examples
mesh = examples.load_airplane()
mesh.cell_points(0)  
```
returns arrays containing the locations of the points which form the cell with the specified cell id.  I can get these point ids using the find_closest_point opperation, but this feels incredibly hacky and I am sure that there is a better way

```
from pyvista import examples
mesh = examples.load_airplane()
points=[mesh.find_closest_point(point) for point in mesh.cell_points(0)]
```

Thank You!

**Comment by MatthewFlamm** [comment, marked as answer]
I think this should be implemented natively in PyVista, but here is some vtk code you can use for this.  This prints the four point ids associated with cell `10`.

```py
import pyvista
mesh = pyvista.Plane()
cell = mesh.GetCell(10)
point_ids = cell.GetPointIds()
points = [point_ids.GetId(i) for i in range(point_ids.GetNumberOfIds())]
print(points)
```

gives

```txt
[11, 12, 23, 22]
```

**Reply by banesullivan** [reply]
Perhaps we should wrap the `vtkCell` class in PyVista and provide `point_ids` as a property that generates that list/array

**Reply by MatthewFlamm** [reply]
I would welcome this.  I'm not sure if we also need a more fully featured implementation of `vtkCellArray`

A first pass would be to give read-only access to the data.  Then we could access with something like `dataset.cells[10].point_ids`.  A significant fraction is already implemented ad-hoc like `dataset.cell_point_ids`.

**Reply by banesullivan** [reply]
> A significant fraction is already implemented ad-hoc like dataset.cell_point_ids.

huh? where is that?

We could also just add a method to `DataSet` like:

```py
def get_cell_point_ids(cell_id):
    cell = mesh.GetCell(10)
    point_ids = cell.GetPointIds()
    return [point_ids.GetId(i) for i in range(point_ids.GetNumberOfIds())]
```

**Reply by MatthewFlamm** [reply]
> huh? where is that?

Sorry I was being lazy with my description and examples.  I was trying to say that existing methods exist for getting info from a specific cell using methods like `DataSet.cell_points` or, a not currently existing, `DataSet.cell_point_ids`.  These are 'ad-hoc' in that each one is a separate implementation.  I was taking from your suggestion of `vtkCell` that we might instead have a property, for example, like `DataSet.cells` that returns a list, although I'm not convinced this is a good idea to return a list, of `vtkCell`s.  Then provide this information from this object.

**Reply by MatthewFlamm** [reply]
The ad-hoc method is implemented in #2897 .  Hopefully will be available in the next release (or the one after).
